### ISSUE TITLE
Crypto++ 5.6.3: CryptoPP::Inflator::UnexpectedEndErr for https://reseed.i2p-projekt.de/

### ISSUE THREAD

**Issue by anonimal** [monero-project/kovri#141]
**Edit:** tl;dr: this is a crypto++ bug. See https://github.com/monero-project/kovri/issues/141#issuecomment-189524508

----
This **only** occurs when reseeding from https://reseed.i2p-projekt.de/ (other reseed servers work without issue)

```
2016:02:21|13:23:33.175623 default:default              NFO             Starting NetDB...
2016:02:21|13:23:33.176412 default:default              NFO             Creating /home/kovri/.kovri/netDb
2016:02:21|13:23:33.179498 default:default              NFO             0 routers loaded
2016:02:21|13:23:33.179784 default:default              NFO             0 floodfills loaded
2016:02:21|13:23:33.181480 default:default              NFO             8 certificates loaded
2016:02:21|13:23:33.181731 default:default              NFO             Downloading SU3 from https://reseed.i2p-projekt.de/
2016:02:21|13:23:34.041831 default:default              NFO             Connected to reseed.i2p-projekt.de:443
2016:02:21|13:23:34.659613 default:default              DBG             Processing file routerInfo-xppk08nufaAidPgZnCo5c4dwvk05WPGECVsGePtHEUE=.dat 652 bytes
terminate called after throwing an instance of 'CryptoPP::Inflator::UnexpectedEndErr'
  what():  Inflator: unexpected end of compressed block
Aborted (core dumped)
```
```
(gdb) bt
#0  0x00007ffff506a3c8 in raise () from /usr/lib/libc.so.6
#1  0x00007ffff506b84a in abort () from /usr/lib/libc.so.6
#2  0x00007ffff5b9bacd in __gnu_cxx::__verbose_terminate_handler () at /build/gcc-multilib/src/gcc-5-20160209/libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007ffff5b99936 in __cxxabiv1::__terminate (handler=<optimized out>) at /build/gcc-multilib/src/gcc-5-20160209/libstdc++-v3/libsupc++/eh_terminate.cc:47
#4  0x00007ffff5b99981 in std::terminate () at /build/gcc-multilib/src/gcc-5-20160209/libstdc++-v3/libsupc++/eh_terminate.cc:57
#5  0x00007ffff5b99b98 in __cxxabiv1::__cxa_throw (obj=0xed8de0, tinfo=0x7ffff6412880 <typeinfo for CryptoPP::Inflator::UnexpectedEndErr>, dest=0x7ffff611b8f0 <CryptoPP::Inflator::UnexpectedEndErr::~UnexpectedEndErr()>)
    at /build/gcc-multilib/src/gcc-5-20160209/libstdc++-v3/libsupc++/eh_throw.cc:87
#6  0x00007ffff611b6e3 in CryptoPP::Inflator::Put2(unsigned char const*, unsigned long, int, bool) () from /usr/lib/libcryptopp.so
#7  0x00000000009bd9e2 in CryptoPP::BufferedTransformation::MessageEnd (this=0x7fffffffdea0, propagation=-1, blocking=true) at /usr/include/cryptopp/cryptlib.h:1328
#8  0x0000000000a52ca9 in i2p::data::Reseeder::ProcessSU3Stream (this=0xed9b80, s=...) at /home/kovri/src/core/Reseed.cpp:254
#9  0x0000000000a521e0 in i2p::data::Reseeder::ReseedFromSU3 (this=0xed9b80, host="https://reseed.i2p-projekt.de/") at /home/kovri/src/core/Reseed.cpp:96
#10 0x0000000000a52104 in i2p::data::Reseeder::ReseedNowSU3 (this=0xed9b80) at /home/kovri/src/core/Reseed.cpp:86
#11 0x0000000000a362fa in i2p::data::NetDb::Reseed (this=0xe94e00 <i2p::data::netdb>) at /home/kovri/src/core/NetworkDatabase.cpp:312
#12 0x0000000000a3521e in i2p::data::NetDb::Start (this=0xe94e00 <i2p::data::netdb>) at /home/kovri/src/core/NetworkDatabase.cpp:79
#13 0x0000000000941733 in i2p::util::Daemon_Singleton::Start (this=0xe943c0 <i2p::util::DaemonLinux::Instance()::instance>) at /home/kovri/src/app/Daemon.cpp:121
#14 0x0000000000967275 in i2p::util::DaemonLinux::Start (this=0xe943c0 <i2p::util::DaemonLinux::Instance()::instance>) at /home/kovri/src/app/DaemonLinux.cpp:127
#15 0x00000000009402aa in main (argc=1, argv=0x7fffffffe8f8) at /home/kovri/src/app/Main.cpp:46
(gdb)
```
```
(gdb) thread apply all bt full
...
Thread 1 (Thread 0x7ffff7fca740 (LWP 5330)):
#0  0x00007ffff506a3c8 in raise () from /usr/lib/libc.so.6
No symbol table info available.
#1  0x00007ffff506b937 in abort () from /usr/lib/libc.so.6
No symbol table info available.
#2  0x00007ffff5b9bacd in __gnu_cxx::__verbose_terminate_handler () at /build/gcc-multilib/src/gcc-5-20160209/libstdc++-v3/libsupc++/vterminate.cc:95
        terminating = true
        t = <optimized out>
#3  0x00007ffff5b99936 in __cxxabiv1::__terminate (handler=<optimized out>) at /build/gcc-multilib/src/gcc-5-20160209/libstdc++-v3/libsupc++/eh_terminate.cc:47
No locals.
#4  0x00007ffff5b99981 in std::terminate () at /build/gcc-multilib/src/gcc-5-20160209/libstdc++-v3/libsupc++/eh_terminate.cc:57
No locals.
#5  0x00007ffff5b99b98 in __cxxabiv1::__cxa_throw (obj=0xed8de0, tinfo=0x7ffff6412880 <typeinfo for CryptoPP::Inflator::UnexpectedEndErr>, dest=0x7ffff611b8f0 <CryptoPP::Inflator::UnexpectedEndErr::~UnexpectedEndErr()>)
    at /build/gcc-multilib/src/gcc-5-20160209/libstdc++-v3/libsupc++/eh_throw.cc:87
        globals = <optimized out>
#6  0x00007ffff611b6e3 in CryptoPP::Inflator::Put2(unsigned char const*, unsigned long, int, bool) () from /usr/lib/libcryptopp.so
No symbol table info available.
#7  0x00000000009bd9e2 in CryptoPP::BufferedTransformation::MessageEnd (this=0x7fffffffdea0, propagation=-1, blocking=true) at /usr/include/cryptopp/cryptlib.h:1328
No locals.
#8  0x0000000000a52ca9 in i2p::data::Reseeder::ProcessSU3Stream (this=0xed9b80, s=...) at /home/kovri/src/core/Reseed.cpp:254
        decompressor = <incomplete type>
        bitFlag = 0
        compressionMethod = 8
        compressedSize = 883
        crc32 = "X\242l&"
        extraFieldLength = 28
        uncompressedSize = 913
        fileNameLength = 59
        localFileName = "routerInfo-xH0gF51WVpsrunJqT0gsGL2twjIP5XejNTwW2M-SU~s=.dat\000\000\000\000\200\201b\327\000\000routerInfo-y5sg1mt58YxBgD0HjZWZYelZngZ3k3pukVCNEBO8pa0=.datUT\005\000\003\327-\311Vux\v\000\001\004\350\003\000\000\004\350\003\000\000PK\001\002\036\003\024\000\000\000\b\000B\037UH^\301-m\211\002\000\000\243\002\000\000;\000\030\000\000\000\000\000\000\000\000\000\200\201_\332\000\000ro"...
        compressed = 0xf00b30 "ۗ(`\260\355\327\216\003k?\337\344n\362X\304ථ݇\321|\356\342\067\227\356\035\324\365\322V\371\064\377ƥ\305\033\256\251(VK-\361j[v;\353\364\236+}-\021\263\"J\213.ʹ\261q/՟\374`\377M\356\217Yϕx\336V\264\035\316(\330x\252Z\274eZ\263\327\217\247+g[fo\272\352\065a\265\277\365\331\302\315\a߳\327o\364;\264\371ddo\372\354\033\027\027\277J\362e\352X-\314\364\347\347\316\320)\177\214\204CV웽G\364dғ^\331{\274/\322y\035\352Om\312\337|\344\246\363\---Type <return> to continue, or q <return> to quit---
323\027|\f\377>\363h\251(\256;\245\350Ǣv}\203nO\307\311\035", <incomplete sequence \351>...
        signature = 67324752
        end = 140737316390630
        magicNumber = "I2Psu3"
        signatureType = 6
        signatureLength = 512
        versionLength = 16 '\020'
        signerIDLength = 16 '\020'
        contentLength = 65463
        fileType = 0 '\000'
        contentType = 3 '\003'
        signerID = "echelon@mail.i2p\000\340\377\377\377\177\000\000\003\000\000\000\000\000\000\000\064\064\063\000\004\000\000\000\004\000\000\000\377\177\000\001\020\063\355", '\000' <repeats 21 times>, "\030\000\000\000\000\000\000\000P\225\355\000\000\000\000\000\025\000\000\000\000\000\000\000\025\000\000\000\000\000\000\000\220i\355\000\000\000\000\000\020\213\355\000\000\000\000\000\061\000\000\000\000\000\000\000\061\000\000\000\000\000\000\000\347\n\224\000\000\000\000\000\002\000\001\273Q\a\a\004\263\262\271\365\377\177", '\000' <repeats 19 times>, "\342\355\000\000\000\000\000\025\000\000\000\006\000\000\000\200\302\350\365\377\177\000\000"...
        it = {first = "echelon@mail.i2p", second = {{
              m_Buf = "\231\301 /\f!̳^\\\263\214J\270p\a\310V\244\350_\216\216,\351/\360\214\033c\215G<\234\215\303\322\024\247\334v6\352\235+\275\313\315\021g\261=JOP\271\202{\021\244S\216\267\236\356ψS\215\210\344H\325\342@\211AG\211\264bh\016\326n\214\250\337\312N\357r\242\257\b\306Ӷ>Q\255#\247\341\322\373\201P\374\217\375\037\366ϦT^L\345?\352\031\221n\004:\365O'\023\257\a͑\320\304\376\274\311Rv\254\361\322\\\303\341\027\204u\207A\265\326`\200O\344\340N\311\360\034х\357\265\307&v8\374\316\036\331\310\062pFcf\214\067\350-\355\332cCVC \027_PA\346\375Zb\321\315\021'\336z"..., ll = {12955766564252074393, 536130985984744542, 3210660126937863880, 5155886118388903913, 15899699904911350844,
                14829154193078105718, 13353260075434731281, 11436766268616702850, 5252473205994082286, 13009007345437565653, 16116285674210879586, 14269848531571265226, 16260004206943844051, 2305156898128788434, 4604170162087055350,
                5761575088999700970, 14181995536372929319, 15200119841547205886, 4721871944518058844, 5683793759848945333, 14390671542350311625, 14472632751875454502, 16732997311898480690, 2324796680091397421, 7087256129000988439,
                341564241306832337, 12692822872924981146, 17200264662140535506, 14754352913824662740, 5124672747427217069, 10146510510545741562, 16266451325545608099, 9102107730512143380, 2030859671930984921, 11521566749808180851,
                2283750992541235096, 10062793148109164492, 5375714042258662217, 8750700772766827991, 6061595500174722898, 8883499985431092483, 6930361492720654388, 17805864981103552542, 5711988316323963243, 5213490682594388285,
                15492718037080431627, 6464093763610794973, 13957807038830998383, 4997933732585168640, 10868072829435861685, 15488975445969926701, 4478074605322080709, 8546239504432181106, 4982544853058825691, 3104862897759053483,
                158103963159778538, 16679501646422244550, 2485237990285635586, 13603549318882451611, 358768611096499512, 2031743118902912617, 12227571951789257866, 17946525382405019570, 13946981461054995852}}}}
        numFiles = 0
        contentPos = 72
#9  0x0000000000a521e0 in i2p::data::Reseeder::ReseedFromSU3 (this=0xed9b80, host="https://reseed.i2p-projekt.de/") at /home/kovri/src/core/Reseed.cpp:96
        ss = {<std::basic_iostream<char, std::char_traits<char> >> = {<std::basic_istream<char, std::char_traits<char> >> = {<std::basic_ios<char, std::char_traits<char> >> = {<std::ios_base> = {
                  _vptr.ios_base = 0x7ffff5e85eb0 <vtable for std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >+104>, static boolalpha = std::_S_boolalpha, static dec = std::_S_dec,
                  static fixed = std::_S_fixed, static hex = std::_S_hex, static internal = std::_S_internal, static left = std::_S_left, static oct = std::_S_oct, static right = std::_S_right, static scientific = std::_S_scientific,
                  static showbase = std::_S_showbase, static showpoint = std::_S_showpoint, static showpos = std::_S_showpos, static skipws = std::_S_skipws, static unitbuf = std::_S_unitbuf, static uppercase = std::_S_uppercase,
                  static adjustfield = std::_S_adjustfield, static basefield = std::_S_basefield, static floatfield = std::_S_floatfield, static badbit = std::_S_badbit, static eofbit = std::_S_eofbit, static failbit = std::_S_failbit,
                  static goodbit = std::_S_goodbit, static app = std::_S_app, static ate = std::_S_ate, static binary = std::_S_bin, static in = std::_S_in, static out = std::_S_out, static trunc = std::_S_trunc,
                  static beg = std::_S_beg, static cur = std::_S_cur, static end = std::_S_end, _M_precision = 6, _M_width = 0, _M_flags = 4098, _M_exception = std::_S_goodbit, _M_streambuf_state = std::_S_goodbit, _M_callbacks = 0x0,
                  _M_word_zero = {_M_pword = 0x0, _M_iword = 0}, _M_local_word = {{_M_pword = 0x0, _M_iword = 0}, {_M_pword = 0x0, _M_iword = 0}, {_M_pword = 0x0, _M_iword = 0}, {_M_pword = 0x0, _M_iword = 0}, {_M_pword = 0x0,
                      _M_iword = 0}, {_M_pword = 0x0, _M_iword = 0}, {_M_pword = 0x0, _M_iword = 0}, {_M_pword = 0x0, _M_iword = 0}}, _M_word_size = 8, _M_word = 0x7fffffffe360, _M_ios_locale = {static none = 0, static ctype = 1,
                    static numeric = 2, static collate = 4, static time = 8, static monetary = 16, static messages = 32, static all = 63, _M_impl = 0x7ffff5e8c860 <(anonymous namespace)::c_locale_impl>,
                    static _S_classic = 0x7ffff5e8c860 <(anonymous namespace)::c_locale_impl>, static _S_global = 0x7ffff5e8c860 <(anonymous namespace)::c_locale_impl>, static _S_categories = 0x7ffff5e7f1e0 <__gnu_cxx::category_names>,
                    static _S_once = 2, static _S_twinned_facets = 0x7ffff5e807c0 <std::locale::_S_twinned_facets>}}, _M_tie = 0x0, _M_fill = 0 '\000', _M_fill_init = false, _M_streambuf = 0x7fffffffe2b8,
                _M_ctype = 0x7ffff5e8c280 <(anonymous namespace)::ctype_c>, _M_num_put = 0x7ffff5e8c210 <(anonymous namespace)::num_put_c>, _M_num_get = 0x7ffff5e8c220 <(anonymous namespace)::num_get_c>},
              _vptr.basic_istream = 0x7ffff5e85e60 <vtable for std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >+24>, _M_gcount = 883}, <std::basic_ostream<char, std::char_traits<char> >> = {
              _vptr.basic_ostream = 0x7ffff5e85e88 <vtable for std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >+64>}, <No data fields>},
          _M_stringbuf = {<std::basic_streambuf<char, std::char_traits<char> >> = {_vptr.basic_streambuf = 0x7ffff5e85af0 <vtable for std::__cxx11::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >+16>,
              _M_in_beg = 0xee8ae0 "I2Psu3", _M_in_cur = 0xee8f10 "PK\003\004\024", _M_in_end = 0xef8cdf "", _M_out_beg = 0xee8ae0 "I2Psu3", _M_out_cur = 0xee8ae0 "I2Psu3", _M_out_end = 0xef8cdf "", _M_buf_locale = {static none = 0,
                static ctype = 1, static numeric = 2, static collate = 4, static time = 8, static monetary = 16, static messages = 32, static all = 63, _M_impl = 0x7ffff5e8c860 <(anonymous namespace)::c_locale_impl>,
                static _S_classic = 0x7ffff5e8c860 <(anonymous namespace)::c_locale_impl>, static _S_global = 0x7ffff5e8c860 <(anonymous namespace)::c_locale_impl>, static _S_categories = 0x7ffff5e7f1e0 <__gnu_cxx::category_names>,
                static _S_once = 2, static _S_twinned_facets = 0x7ffff5e807c0 <std::locale::_S_twinned_facets>}}, _M_mode = 24,
            _M_string = "I2Psu3\000\000\000\006\002\000\000\020\000\020\000\000\000\000\000\000\377\267\000\000\000\003", '\000' <repeats 12 times>, "\061\064\065\066\060\062\065\062\064\070\000\000\000\000\000\000echelon@mail.i2pPK\003\004\024\000\000\000\b\000\004 UHX\242l&s\003\000\000\221\003\000\000;\000\034\000routerInfo-xH0gF51WVpsrunJqT0gsGL2twjIP5XejNTwW2M-SU~s=.datUT\t\000\003\067(\311V\212.\311Vux\v\000\001\004\350\003\000\000\004\350\003\000\000ۗ(`\260\355\327\216\003k?"...}}
        url = "https://reseed.i2p-projekt.de/i2pseeds.su3"
        su3 = "I2Psu3\000\000\000\006\002\000\000\020\000\020\000\000\000\000\000\000\377\267\000\000\000\003", '\000' <repeats 12 times>, "\061\064\065\066\060\062\065\062\064\070\000\000\000\000\000\000echelon@mail.i2pPK\003\004\024\000\000\000\b\000\004 UHX\242l&s\003\000\000\221\003\000\000;\000\034\000routerInfo-xH0gF51WVpsrunJqT0gsGL2twjIP5XejNTwW2M-SU~s=.datUT\t\000\003\067(\311V\212.\311Vux\v\000\001\004\350\003\000\000\004\350\003\000\000ۗ(`\260\355\327\216\003k?"...
#10 0x0000000000a52104 in i2p::data::Reseeder::ReseedNowSU3 (this=0xed9b80) at /home/kovri/src/core/Reseed.cpp:86
        s = 6
        ind = 1
        reseedHost = "https://reseed.i2p-projekt.de/"
#11 0x0000000000a362fa in i2p::data::NetDb::Reseed (this=0xe94e00 <i2p::data::netdb>) at /home/kovri/src/core/NetworkDatabase.cpp:312
        result = 10521650
        reseedRetries = 0
#12 0x0000000000a3521e in i2p::data::NetDb::Start (this=0xe94e00 <i2p::data::netdb>) at /home/kovri/src/core/NetworkDatabase.cpp:79
No locals.
#13 0x0000000000941733 in i2p::util::Daemon_Singleton::Start (this=0xe943c0 <i2p::util::DaemonLinux::Instance()::instance>) at /home/kovri/src/app/Daemon.cpp:121
No locals.
#14 0x0000000000967275 in i2p::util::DaemonLinux::Start (this=0xe943c0 <i2p::util::DaemonLinux::Instance()::instance>) at /home/kovri/src/app/DaemonLinux.cpp:127
        pid = "5330\n\000\000\000\002"
        sa = {__sigaction_handler = {sa_handler = 0x966e15 <handle_signal(int)>, sa_sigaction = 0x966e15 <handle_signal(int)>}, sa_mask = {__val = {0 <repeats 16 times>}}, sa_flags = 268435456,
---Type <return> to continue, or q <return> to quit---
          sa_restorer = 0xe96580 <i2p::util::log::g_Log>}
#15 0x00000000009402aa in main (argc=1, argv=0x7fffffffe8f8) at /home/kovri/src/app/Main.cpp:46
No locals.
(gdb)
```
More on the way. Will update ticket as information comes in.

All-in-all, our SU3 processing code needs a rewrite regardless of how this issue is resolved.

**Comment by anonimal** [comment]
May be a server-side bug. Awaiting response.
```
2016-02-22 06:53:02     &anonimal       zzz: So, I think echelon's reseed is broken.
2016-02-22 06:53:41     +anonimal       eche|on: Hi, can you please confirm that https://reseed.i2p-projekt.de/ is serving correct SU3's?
2016-02-22 06:54:48     +anonimal       $ wget -S -c --no-check-certificate https://reseed.i2p-projekt.de/i2pseeds.su3  # <-- confirms that 0 bytes are returned yet response is 200 OK
```

**Comment by anonimal** [comment]
Not the easy-bug I thought it would be. More research needed.
```
2016-02-23 12:31:08     +anonimal       eche|off: Did you get my message yesterday about reseed.i2p-projekt.de?
2016-02-23 12:32:09     eche|off        no, why?
2016-02-23 12:32:34     +anonimal       eche|off: I'll repaste
2016-02-23 12:35:18     +anonimal       2016-02-22 06:53:41     +anonimal       eche|on: Hi, can you please confirm that https://reseed.i2p-projekt.de/ is serving correct SU3's?
2016-02-23 12:36:13     eche|off        [23/Feb/2016:13:35:52 +0100] "GET http://reseed.i2p-projekt.de/ HTTP/1.1" 200 326 "-" "Wget/1.11.4"
2016-02-23 12:36:21     eche|off        so it works ;-)
2016-02-23 12:36:28     eche|off        [23/Feb/2016:13:35:40 +0100] "GET /i2pseeds.su3 HTTP/1.1" 200 22851 "-" "Wget/1.11.4"
2016-02-23 12:36:33     +anonimal       eche|off: No, look at i2pseeds.su3
2016-02-23 12:36:36     +anonimal       It's 0 bytes
2016-02-23 12:36:43     +anonimal       Other reseeds actually send an SU3
2016-02-23 12:36:45     eche|off        it just does not send out to some wandom wgets
2016-02-23 12:37:36     eche|off        [23/Feb/2016:13:37:24 +0100] "GET /i2pseeds.su3 HTTP/1.1" 200 91345 "-" "Wget/1.11.4"
2016-02-23 12:37:42     eche|off        the server is strctly locked down
2016-02-23 12:38:44     +anonimal       eche|off: Ah, I change UA to that Wget version. Thanks.
```

**Comment by anonimal** [comment]
```
2016-02-23 14:16:55     &anonimal       So, I think I figured out #141.
2016-02-23 14:17:25     &anonimal       For some reason, only echelon mixes compressed and uncompressed .dat's in .su3's.
2016-02-23 14:19:08     &anonimal       So, when we come upon an uncompressed .dat *after* a compressed .dat, we don't extract it from the container but instead blow up.
2016-02-23 14:43:59     &anonimal       My choice of wording wasn't clear. I mean deflated .dat's versus contained .dat's.
2016-02-23 14:44:12      *      anonimal talking more in #i2p-dev
2016-02-23 14:54:48     &anonimal       So, it's confirmed: reseed code is ugly and broken and needs to move up on the priority list; in particular our deflating/inflating impl.
2016-02-23 14:54:55     &anonimal       I'll add a note in ticket.
```
```
$ unzip -vl reseed.i2p-projekt.de_i2pseeds.su3.extracted
Archive:  reseed.i2p-projekt.de_i2pseeds.su3.extracted
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
     882  Defl:N      718  19% 2016-02-21 03:09 3c61df75  routerInfo-wwSSvi2Dms8soraxsPs4a0o5RhPKlVyQ3eSwG2RhEjw=.dat
     956  Defl:N      840  12% 2016-02-21 02:52 d59e0974  routerInfo-FflLebcHF3N8ZFJ2AbOQsfvACXcSz7-XNl84-p6jn5c=.dat
     914  Defl:N      883   3% 2016-02-21 03:03 62933b3a  routerInfo-6mKKMRSNbAaqS2lg-F2sfDU5A19ybLGEa5iy0QCLhEU=.dat
    1195  Defl:N     1059  11% 2016-02-21 02:38 0940934a  routerInfo-62640gnXGCF5TiLZ21SV2UNt2BfODYlrpTJvGWybRFk=.dat
     551  Stored      551   0% 2016-02-21 03:04 6e31a912  routerInfo-JgPQm8tMPtPpMf3NZNpSf339mcodfDXuw7IC-M1podE=.dat
    1026  Defl:N      873  15% 2016-02-21 02:32 797cfbe2  routerInfo-CYC~XTXOMntBVLpgiPed54l4ZJD-ZTFaOGDnDcOU1sA=.dat

$ unzip reseed.i2p-projekt.de_i2pseeds.su3.extracted
Archive:  reseed.i2p-projekt.de_i2pseeds.su3.extracted
  inflating: routerInfo-wwSSvi2Dms8soraxsPs4a0o5RhPKlVyQ3eSwG2RhEjw=.dat
  inflating: routerInfo-FflLebcHF3N8ZFJ2AbOQsfvACXcSz7-XNl84-p6jn5c=.dat
  inflating: routerInfo-6mKKMRSNbAaqS2lg-F2sfDU5A19ybLGEa5iy0QCLhEU=.dat
  inflating: routerInfo-62640gnXGCF5TiLZ21SV2UNt2BfODYlrpTJvGWybRFk=.dat
 extracting: routerInfo-JgPQm8tMPtPpMf3NZNpSf339mcodfDXuw7IC-M1podE=.dat
  inflating: routerInfo-CYC~XTXOMntBVLpgiPed54l4ZJD-ZTFaOGDnDcOU1sA=.dat

=======================================

$ unzip -vl uk.reseed.i2p2.no_i2pseeds.su3.extracted
Archive:  uk.reseed.i2p2.no_i2pseeds.su3.extracted
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
     967  Stored      967   0% 2016-02-22 00:20 ba6ee8f8  routerInfo-JI3-pNnEr3MZGk3gPxLcSTbG7pSeyfWk6~6LYWA4suY=.dat
     680  Stored      680   0% 2016-02-22 00:20 fd4e3ac0  routerInfo-jxOk0~XRC8h3VF5bwgcAjZazZmDc-43~Cf~4R6IGMA0=.dat
     678  Stored      678   0% 2016-02-22 00:20 7bc916e5  routerInfo-00LteHmEWtf8Gc0saUXrku-DQU4tY2gH1WGqOA1x87c=.dat
     684  Stored      684   0% 2016-02-22 00:20 374bf89e  routerInfo-~LdFhps75hR-TxbyuLKZ7raceAs9izwb4jTWFbZydWA=.dat
     813  Stored      813   0% 2016-02-22 00:20 59110b24  routerInfo-Ruab3S0n-SwHxc2g--vk3303p8z~9PhoLsSCA-bmGsM=.dat

$ unzip uk.reseed.i2p2.no_i2pseeds.su3.extracted
Archive:  uk.reseed.i2p2.no_i2pseeds.su3.extracted
 extracting: routerInfo-JI3-pNnEr3MZGk3gPxLcSTbG7pSeyfWk6~6LYWA4suY=.dat
 extracting: routerInfo-jxOk0~XRC8h3VF5bwgcAjZazZmDc-43~Cf~4R6IGMA0=.dat
 extracting: routerInfo-00LteHmEWtf8Gc0saUXrku-DQU4tY2gH1WGqOA1x87c=.dat
 extracting: routerInfo-~LdFhps75hR-TxbyuLKZ7raceAs9izwb4jTWFbZydWA=.dat
 extracting: routerInfo-Ruab3S0n-SwHxc2g--vk3303p8z~9PhoLsSCA-bmGsM=.dat

=======================================

$ unzip -vl user.mx24.eu_i2pseeds.su3.extracted
Archive:  user.mx24.eu_i2pseeds.su3.extracted
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
     675  Defl:N      657   3% 2016-02-20 07:08 ed20ebc3  routerInfo-OH~CXeIH~Ql1KnuSYMxqgFp8V011SGsUEfKIieYu~v4=.dat
     675  Defl:N      655   3% 2016-02-20 07:11 8264619a  routerInfo-lQq7D-Y-weLlgcNxMzk1yXYzMNAmgyBKLOJbk8PNzxQ=.dat
     712  Defl:N      677   5% 2016-02-20 06:38 9b4f8817  routerInfo-VvRiaeFEXxeG67vH62Cl9imi9cgUQKlczoGfRoYgqOU=.dat
     643  Defl:N      624   3% 2016-02-20 06:56 6225ec5b  routerInfo-PxhfJ11sFBexvRokeVLnPml1CZ6m8MEoaW-zB8TtiJ8=.dat
     729  Defl:N      696   5% 2016-02-20 06:30 ed8603e5  routerInfo-Cvkc510hBcvNn4IcWULT58EEfO0wLq20ZBrOcjJGf8I=.dat

$ unzip user.mx24.eu_i2pseeds.su3.extracted
Archive:  user.mx24.eu_i2pseeds.su3.extracted
  inflating: routerInfo-OH~CXeIH~Ql1KnuSYMxqgFp8V011SGsUEfKIieYu~v4=.dat
  inflating: routerInfo-lQq7D-Y-weLlgcNxMzk1yXYzMNAmgyBKLOJbk8PNzxQ=.dat
  inflating: routerInfo-VvRiaeFEXxeG67vH62Cl9imi9cgUQKlczoGfRoYgqOU=.dat
  inflating: routerInfo-PxhfJ11sFBexvRokeVLnPml1CZ6m8MEoaW-zB8TtiJ8=.dat
  inflating: routerInfo-Cvkc510hBcvNn4IcWULT58EEfO0wLq20ZBrOcjJGf8I=.dat
```

**Comment by anonimal** [comment]
```
2016-02-26 23:02:51      *      anonimal sighs
2016-02-26 23:02:56     &anonimal       Well, that was annoying.
2016-02-26 23:03:25     &anonimal       I ended up patching #141, simple one-liner, but it turns out that the problem was a new (unknown to me) API caveat (or maybe bug?) with crypto++ 5.6.3, of which was only introduced to Arch 11 days ago.
2016-02-26 23:03:39     &anonimal       Of course I can't reproduce #141 on debian or ubuntu; both of which use older crypto++ versions.
2016-02-26 23:04:07     &anonimal       But what baffles me is *why* I would need to add an extra byte to compressedSize for certain RI entries; AFAICT nothing applicable had changed to crypto++'s Inflater class between 5.6.2->3.
2016-02-26 23:04:26     &anonimal       Maybe it's not taking into account a null byte (or any byte) at the end of said RI entries? But net.i2p.data.router.RouterInfo proves that all RI's are intact...
2016-02-26 23:04:39     &anonimal       Anyway, I'm simply speaking out loud. I'll need some more time with this because it doesn't look right and there has to be a better way to do this.
2016-02-26 23:04:46     &anonimal       Will paste this in ticket.
```

**Comment by anonimal** [comment]
Confirmed crypto++ bug (or crypto++ related). Downgrading to <=5.6.2 solves for now.  Will keep this issue open until:

1. I gather enough information for a proper upstream bug report
2. I triple check my existing patch to see if it's safe to implement
3. I find the time to do both 1. and 2.

**Comment by EinMByte** [comment]
If it's a Crypto++ bug, then waiting for it to get fixed could take a long time. There were two years between 5.6.2 and 5.6.3. That said, it would be good to provide them with a bug report or patch.

**Comment by anonimal** [comment]
Helpful news from zzz:
```
2016-02-27 22:27:39     +zzz    [23:26:07] anonimal, there's a similar quirk in java's gzip inflater
2016-02-27 22:27:39     +zzz    [23:26:45] quote ResettableGZIPInputStream.java:
2016-02-27 22:27:39     +zzz    [23:27:10]      *  Why? otherwise the inflater finished() is wrong when the compressed payload
2016-02-27 22:27:39     +zzz    [23:27:10]      *  (in between the 10 byte header and the 8 byte footer) is a multiple of 512 bytes,
2016-02-27 22:27:39     +zzz    [23:27:10]      *  which caused read(buf, off, len) above to fail.
2016-02-27 22:27:39     +zzz    [23:27:10]      *  Happened every time with 1042 byte compressed router infos, for example.
2016-02-27 22:27:39     +zzz    [23:27:35]      *     "Note: When using the 'nowrap' option it is also necessary to provide an extra "dummy" byte as input.
2016-02-27 22:27:39     +zzz    [23:27:35]      *      This is required by the ZLIB native library in order to support certain optimizations."
2016-02-27 22:27:39     +zzz    [23:27:35]      *
2016-02-27 22:27:39     +zzz    [23:27:35]      *     http://docs.oracle.com/javase/1.5.0/docs/api/java/util/zip/Inflater.html
2016-02-27 22:27:39     +zzz    [23:27:56]      *     "The fix simply involves copying the byte array and tacking a single null byte on to the end."
2016-02-27 22:27:39     +zzz    [23:27:56]      *
2016-02-27 22:27:39     +zzz    [23:27:56]      *     http://code.google.com/p/google-apps-sso-sample/issues/detail?id=8
2016-02-27 22:27:39     +zzz    [23:28:16]      *     http://bugs.sun.com/bugdatabase/view_bug.do?bug_id=4795299
2016-02-27 22:27:39     +zzz    [13:46:31] anonimal, hope the inflater stuff I posted was helpful
2016-02-27 22:31:28     &anonimal       zzz: THANKS! I thought I was going crazy, that makes *complete* sense.
```

**Comment by anonimal** [comment]
```
2016-02-29 14:57:33     &anonimal       zzz: Is that a typo in ResettableGZIPInputStream.java line 295? s/1042/1024/ ?
2016-02-29 16:37:39     +zzz    anonimal, no, it's 10 + (n * 512) + 8, = 1042 for n = 2
2016-02-29 17:31:07     &anonimal       Oh, ok. The "multiple of 512 bytes" comment before made the 1042 questionable.
2016-02-29 17:34:49     +zzz    probably written in a rage after i spent a week finding and fixing it :)
2016-02-29 21:45:05     &anonimal       zzz: lol I know that rage. I'd be there now if you didn't confirm the null byte oddity.
2016-02-29 21:45:33     &anonimal       So, #141 is safely patched for now but (10 + (n * 512) + 8) doesn't seem to apply; the problem child is a 720 byte compressed payload.
2016-02-29 21:46:00     &anonimal       This means that I have to tack a null byte at the end of every payload which seems silly, but if I can't find the magic number then they all need to be doused.
2016-02-29 21:46:20     &anonimal       (and if this is a number issue then, should a 719 byte payload come along, inflation should still work because a null byte is a null byte?)
2016-02-29 21:46:55     &anonimal       We'll see, I don't see anything in the zlib spec about our issue, and it doesn't help that crypto++ did a refactor between the two versions from unix to windows carriage returns, so now diffs are all spammy and ^M.
2016-02-29 22:08:51     +zzz    anonimal, we just add the null byte all the time. The java specs don't indicate any particular length as problematic, and I'm sure it could change based on the underlying implementation
2016-02-29 22:40:18     &anonimal       zzz: Excellent, thanks.
```
Referencing 8c61f5d. I'll PR once other reseed work is done.

**Comment by anonimal** [comment]
I referenced this issue in an [already-open cryptopp issue](https://github.com/weidai11/cryptopp/issues/112). This ticket is now resolved and will close once we merge into master.
